jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java

```diff
@@ -152,13 +193,465 @@ default int capacity()
     }
 
     /**
+     * @return the number of bytes left for appending in the {@code ByteBuffer}
+     */
+    default int space()
+    {
+        return capacity() - remaining();
+    }
+
+    /**
+     * @return whether the {@code ByteBuffer} has remaining bytes left for appending
+     */
+    default boolean isFull()
+    {
+        return space() == 0;
+    }
+
+    /**
+     * Clears the contained byte buffer to be empty in flush mode.
      * @see BufferUtil#clear(ByteBuffer)
      */
     default void clear()
     {
         BufferUtil.clear(getByteBuffer());
     }
 
+    /**
+     * Copies the contents of the given byte buffer at the end of this buffer.
+     * @param bytes the byte buffer to copy from.
+     * @return true if all bytes of the given buffer were copied, false otherwise.
+     * @see BufferUtil#append(ByteBuffer, ByteBuffer)
+     */
+    default boolean append(ByteBuffer bytes)
+    {
+        BufferUtil.append(getByteBuffer(), bytes);
+        return !bytes.hasRemaining();
+    }
+
+    /**
+     * Copies the contents of the given retainable byte buffer at the end of this buffer.
+     * @param bytes the retainable byte buffer to copy from.
+     * @return true if all bytes of the given buffer were copied, false otherwise.
+     * @see BufferUtil#append(ByteBuffer, ByteBuffer)
+     */
+    default boolean append(RetainableByteBuffer bytes)
+    {
+        return bytes.remaining() == 0 || append(bytes.getByteBuffer());
+    }
+
+    /**
+     * Copies the contents of this retainable byte buffer at the end of the given byte buffer.
+     * @param toInfillMode the destination buffer.
+     * @see ByteBuffer#put(ByteBuffer)
+     */
+    default void putTo(ByteBuffer toInfillMode)
+    {
+        toInfillMode.put(getByteBuffer());
+    }
+
+    /**
+     * Asynchronously copies the contents of this retainable byte buffer into given sink.
+     * @param sink the destination sink.
+     * @param last true if this is the last write.
+     * @param callback the callback to call upon the write completion.
+     * @see org.eclipse.jetty.io.Content.Sink#write(boolean, ByteBuffer, Callback)
+     */
+    default void writeTo(Content.Sink sink, boolean last, Callback callback)
+    {
+        sink.write(last, getByteBuffer(), callback);
+    }
+
+    /**
+     * An aggregating {@link RetainableByteBuffer} that may grow when content is appended to it.
+     */
+    class Aggregator implements RetainableByteBuffer
```

Comment:
I don't think this is a good idea.

`Aggregator` should _have-a_ RBB, not _be_ a RBB.

Reply:
The whole point of this exercise is to make the RBB API into something that hides the implementation, which might be:
 + a read-only fixed `ByteBuffer` with no space
 + a read/write single `ByteBuffer` with finite space 
 + a read/write single `ByteBuffer` that may grow to a limit
 + a read/write chain of zero copy buffers

Currently we have it that both the appender and the appendee have to agree on the nature of the implementation.    By making RBB able to hide these details, then the appender need not know and only the creator of the appendee has to choose the implementation.   This will not always be perfect, but it is 50% better than before.

See more general discussion below...

Reply:
The other reason that Aggregator should be-a RBB is that forces it to have the same append signature.    We have too many similar by different append/put signatures in the code base: ByteBuffer, BufferUtil, appenders, aggregators, etc.     Having a single API on RBB allows many of these needless API differences to be removed.